Segment fault after module definition when using functions like `puts`

Please answer these questions before submitting your issue. Thanks!

### What version of Go are you using (`go version`)?
go version go1.8.1 darwin/amd64

### What operating system and processor architecture are you using (`go env`)?
```shell
GOARCH="amd64"
GOBIN="/Users/mitch/bin"
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOOS="darwin"
GOPATH="/Users/mitch"
GORACE=""
GOROOT="/usr/local/Cellar/go/1.8.1/libexec"
GOTOOLDIR="/usr/local/Cellar/go/1.8.1/libexec/pkg/tool/darwin_amd64"
GCCGO="gccgo"
CC="clang"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/65/106pnkdj4q158c3xbx9h6ksr0000gn/T/go-build213543851=/tmp/go-build -gno-record-gcc-switches -fno-common"
CXX="clang++"
CGO_ENABLED="1"
PKG_CONFIG="pkg-config"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
```

### What version of GoRuby are you using (`git show -s --format=%H` or goruby -version)?
master (ec0bb1dc93f3edd8253ffedde46ea2f8ebbc8206)

### What did you do?

```shell
$ module_segment_fault=<<EOF
module Foo
end

puts 'hello'
EOF
$ go run main.go -e "${module_segment_fault}"
```

### What did you expect to see?
`hello`

### What did you see instead?
```shell
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x10bc60d]

goroutine 1 [running]:
github.com/goruby/goruby/object.Send(0x117c540, 0xc420086500, 0xc420082590, 0x4, 0xc4200864f0, 0x1, 0x1, 0x1, 0x0, 0x0, ...)
	/Users/mitch/src/github.com/goruby/goruby/object/send.go:6 +0x5d
github.com/goruby/goruby/evaluator.Eval(0x123a0c8, 0xc4200980a0, 0x117d760, 0xc4200824e0, 0x0, 0x0, 0x0, 0x0)
	/Users/mitch/src/github.com/goruby/goruby/evaluator/evaluator.go:104 +0xe33
github.com/goruby/goruby/evaluator.Eval(0x123a068, 0xc42007e750, 0x117d760, 0xc4200824e0, 0x0, 0x0, 0x0, 0x0)
	/Users/mitch/src/github.com/goruby/goruby/evaluator/evaluator.go:26 +0x117d
github.com/goruby/goruby/evaluator.evalProgram(0xc4200825c0, 0x2, 0x2, 0x117d760, 0xc4200824e0, 0xc420082580, 0x0, 0x1d, 0x20)
	/Users/mitch/src/github.com/goruby/goruby/evaluator/evaluator.go:147 +0x95
github.com/goruby/goruby/evaluator.Eval(0x117ba40, 0xc4200825a0, 0x117d760, 0xc4200824e0, 0x0, 0x0, 0x0, 0x0)
	/Users/mitch/src/github.com/goruby/goruby/evaluator/evaluator.go:24 +0x278
github.com/goruby/goruby/interpreter.(*interpreter).Interpret(0xc420086260, 0xc420082580, 0x1d, 0x21d, 0xc420082580, 0x1d, 0x0)
	/Users/mitch/src/github.com/goruby/goruby/interpreter/interpreter.go:32 +0xc3
main.main()
	/Users/mitch/src/github.com/goruby/goruby/main.go:53 +0x220
exit status 2
```

--- a/evaluator/evaluator.go
+++ b/evaluator/evaluator.go
@@ -77,13 +77,13 @@ func Eval(node ast.Node, env object.Environment) (object.RubyObject, error) {
 		return val, nil
 	case *ast.ModuleExpression:
 		module := object.NewModule(node.Name.Value, nil)
-		env.Set("self", &object.Self{module, node.Name.Value})
-		defer env.Unset("self")
-		bodyReturn, err := Eval(node.Body, env)
+		moduleEnv := object.NewEnclosedEnvironment(env)
+		moduleEnv.Set("self", &object.Self{module, node.Name.Value})
+		bodyReturn, err := Eval(node.Body, moduleEnv)
 		if err != nil {
 			return nil, err
 		}
-		selfObject, _ := env.Get("self")
+		selfObject, _ := moduleEnv.Get("self")
 		self := selfObject.(*object.Self)
 		env.SetGlobal(node.Name.Value, self.RubyObject)
 		return bodyReturn, nil

--- a/evaluator/evaluator_test.go
+++ b/evaluator/evaluator_test.go
@@ -254,75 +254,107 @@ func TestVariableAssignmentExpression(t *testing.T) {
 }
 
 func TestModuleObject(t *testing.T) {
-	tests := []struct {
-		input           string
-		expectedName    string
-		expectedMethods map[string]string
-		expectedReturn  object.RubyObject
-	}{
-		{
-			`module Foo
+	t.Run("module definition", func(t *testing.T) {
+		tests := []struct {
+			input           string
+			expectedName    string
+			expectedMethods map[string]string
+			expectedReturn  object.RubyObject
+		}{
+			{
+				`module Foo
 				def a
-					"foo"
+				"foo"
 				end
-			end`,
-			"Foo",
-			map[string]string{"a": "fn() {\nfoo\n}"},
-			&object.Symbol{"a"},
-		},
-		{
-			`module Foo
+				end`,
+				"Foo",
+				map[string]string{"a": "fn() {\nfoo\n}"},
+				&object.Symbol{"a"},
+			},
+			{
+				`module Foo
 				3
-			end`,
-			"Foo",
-			map[string]string{},
-			&object.Integer{3},
-		},
-		{
-			`module Foo
-			end`,
-			"Foo",
-			map[string]string{},
-			object.NIL,
-		},
-	}
-
-	for _, tt := range tests {
-		env := object.NewEnvironment()
-		env.Set("self", &object.Self{&object.Object{}, "main"})
-		evaluated, err := testEval(tt.input, env)
-		checkError(t, err)
-
-		if !reflect.DeepEqual(evaluated, tt.expectedReturn) {
-			t.Logf("Expected return object to equal\n%+#v\n\tgot\n%+#v\n", tt.expectedReturn, evaluated)
-			t.Fail()
+				end`,
+				"Foo",
+				map[string]string{},
+				&object.Integer{3},
+			},
+			{
+				`module Foo
+				end`,
+				"Foo",
+				map[string]string{},
+				object.NIL,
+			},
 		}
 
-		module, ok := env.Get(tt.expectedName)
+		for _, tt := range tests {
+			env := object.NewEnvironment()
+			env.Set("self", &object.Self{&object.Object{}, "main"})
+			evaluated, err := testEval(tt.input, env)
+			checkError(t, err)
+
+			if !reflect.DeepEqual(evaluated, tt.expectedReturn) {
+				t.Logf("Expected return object to equal\n%+#v\n\tgot\n%+#v\n", tt.expectedReturn, evaluated)
+				t.Fail()
+			}
+
+			module, ok := env.Get(tt.expectedName)
+			if !ok {
+				t.Logf("Expected module to exist in env")
+				t.Logf("Env: %+#v\n", env)
+				t.FailNow()
+			}
+
+			actualMethods := make(map[string]string)
+
+			methods := module.Class().Methods().GetAll()
+			for name, method := range methods {
+				if function, ok := method.(*object.Function); ok {
+					actualMethods[name] = function.Inspect()
+				}
+			}
+
+			if !reflect.DeepEqual(tt.expectedMethods, actualMethods) {
+				t.Logf(
+					"Expected module methods to equal\n%+#v\n\tgot\n%+#v\n",
+					tt.expectedMethods,
+					actualMethods,
+				)
+				t.Fail()
+			}
+		}
+	})
+	t.Run("self after module definition", func(t *testing.T) {
+		input := `
+		module Foo
+			def a
+				"foo"
+			end
+		end
+		`
+
+		main := &object.Self{&object.Object{}, "main"}
+		env := object.NewEnvironment()
+		env.Set("self", main)
+		_, err := testEval(input, env)
+		checkError(t, err)
+
+		self, ok := env.Get("self")
 		if !ok {
-			t.Logf("Expected module to exist in env")
-			t.Logf("Env: %+#v\n", env)
+			t.Logf("Expected self in the env")
 			t.FailNow()
 		}
 
-		actualMethods := make(map[string]string)
-
-		methods := module.Class().Methods().GetAll()
-		for name, method := range methods {
-			if function, ok := method.(*object.Function); ok {
-				actualMethods[name] = function.Inspect()
-			}
-		}
-
-		if !reflect.DeepEqual(tt.expectedMethods, actualMethods) {
+		if !reflect.DeepEqual(main, self) {
 			t.Logf(
-				"Expected module methods to equal\n%+#v\n\tgot\n%+#v\n",
-				tt.expectedMethods,
-				actualMethods,
+				"Expected self to equal\n%+#v\n\tgot\n%+#v\n",
+				main,
+				self,
 			)
 			t.Fail()
 		}
-	}
+	})
 }
 
 func TestFunctionObject(t *testing.T) {

--- a/interpreter/interpreter_test.go
+++ b/interpreter/interpreter_test.go
@@ -1,6 +1,7 @@
 package interpreter
 
 import (
+	"reflect"
 	"testing"
 
 	"github.com/goruby/goruby/object"
@@ -72,4 +73,57 @@ func TestInterpreterInterpret(t *testing.T) {
 			t.Fail()
 		}
 	})
+}
+
+func TestSelfAfterModuleDefinition(t *testing.T) {
+	input := `
+		module Foo
+		end
+
+		self
+	`
+	interpreter := New()
+
+	defer func() {
+		if r := recover(); r != nil {
+			t.Logf("Expected no panic, got %T:%v\n", r, r)
+			t.Fail()
+		}
+	}()
+	evaluated, err := interpreter.Interpret(input)
+
+	if err != nil {
+		t.Logf("Expected no error, got %T:%v", err, err)
+		t.Fail()
+	}
+
+	expected := &object.Self{&object.Object{}, "main"}
+	if !reflect.DeepEqual(expected, evaluated) {
+		t.Logf("Expected self to equal\n%+#v\n\tgot\n%+#v\n", expected, evaluated)
+		t.Fail()
+	}
+}
+
+func TestModuleInEnv(t *testing.T) {
+	input := `
+		module Foo
+			def foo
+				3
+			end
+		end
+		Foo
+	`
+	interpreter := New()
+
+	evaluated, err := interpreter.Interpret(input)
+	if err != nil {
+		t.Logf("Expected no error, got %T:%v", err, err)
+		t.Fail()
+	}
+
+	_, ok := evaluated.(*object.Module)
+	if !ok {
+		t.Logf("Expected evaluated return to be a object.Module, got %T", evaluated)
+		t.Fail()
+	}
 }

